docs/src/explanations/optional-groups.md

```diff
@@ -0,0 +1,177 @@
+      }
+    }
+    group(A.D) {
+      val d = WireInit(port ^ a)
+    }
+  }
+}
+```
+
+## Conventions
+
+Currently, there is only one supported convetion, `Bind`.  This will cause
+optional group definitions to be lowered to Verilog modules that are
+instantiated via the SystemVerilog `bind` mechanism.  The lowering to Verilog of
+nested groups avoids illegal nested usage of `bind`.
+
+More conventions may be supported in the future.
+
+## Examples
+
+### Design Verification Example
+
+Consider a use case where a design or design verification engineer would like to
+add some asserts and debug prints to a module.  The logic necessary for the
+asserts and debug prints requires additional computation.  All of this code
+should not be included in the final Verilog.  The engineer can use three
+optional groups to do this.
+
+There are three optional groups that emerge from this example:
+
+1. A common `Verification` group
+1. An `Assert` group nested under the `Verification` group
+1. A `Debug` group also nested under the `Verification` group
+
+The `Verification` group can be used to store common logic used by both the
+`Assert` and `Debug` groups.  The latter two groups allow for separation of,
+respectively, assertions from prints.
+
+One way to write this in Scala is the following:
+
+```scala mdoc:reset:silent
+import chisel3._
+import chisel3.group.{Convention, Declaration}
+
+// All groups are declared here.  The Assert and Debug groups are nested under
+// the Verification group.
+object Verification extends group.Declaration(group.Convention.Bind) {
+  object Assert extends group.Declaration(group.Convention.Bind)
+  object Debug extends group.Declaration(group.Convention.Bind)
+}
+
+class Foo extends Module {
+  val a = IO(Input(UInt(32.W)))
+  val b = IO(Output(UInt(32.W)))
+
+  b := a +% 1.U
+
+  // This defines the `Verification` group inside Foo.
+  group(Verification) {
+
+    // Some common logic added here.  The input port `a` is "captured" and
+    // used here.
+    val a_d0 = RegNext(a)
+
+    // This defines the `Assert` group.
+    group(Verification.Assert) {
+      chisel3.assert(a >= a_d0, "a must always increment")
+    }
+
+    // This defines the `Debug` group.
+    group(Verification.Debug) {
+      printf("a: %x, a_d0: %x", a, a_d0)
+    }
+  }
+
+}
+
+```
+
+After compilation, this will produce four Verilog files (one for `Foo` and one
+for each optional group definition in module `Foo`):
+
+1. `Foo`
+1. `Foo_Verification`
+1. `Foo_Verification_Assert`
+1. `Foo_Verification_Debug`
+
+The ports of each module created from an optional group will be automatically
+determined based on what that group captured from outside the group.  In the
+example above, the `Verification` group definition captured port `a`.  Both the
+`Assert` and `Debug` group definitions captured `a` and `a_d0`.
+
+Additionally, three bind files (one for each optional group declaration):
+
+1. `groups_Foo_Verification.sv`
```

Comment:
are these the bind files? Given the warning below, maybe you should mention something like "The ones prefixed by `groups_`". And I guess I am suprised that these files don't have the name `_bindings` or something similar in the name...

Reply:
Yes, these are files that contain Verilog bind statements.

The warning should be located earlier, during the discussion of the Verilog modules that are produced, not the bind files. The bind files are part of the ABI and must exist. Good thinking.

This could or could not include "_bindings" in the filename. However, that would be better discussed on the FIRRTL spec given that `groups_<circuit>_<root>{_<nested>}*.sv` is the current format (see: https://github.com/chipsalliance/firrtl-spec/blob/main/abi.md#bind-lowering-convention). The main reason to not include `_bindings` is that an SV bind is only one envisioned convention and the file, if included, enables the group. Whether or not this is implemented via SV bind is arguably an implementation. There is expected to eventually be an `ifdef` group convention that will lower groups to inline Verilog `ifdefs`. If we keep the filename as `groups_` then the filename can be the same regardless of the group convention used.

WDYT?